### ISSUE TITLE
[Bug report] "There was a problem finishing installing dependencies" v6.0.0

### ISSUE THREAD

**Issue by alexbarton** [olets/zsh-abbr#156]
@olets's maintainer edit: this has become a catchall for several bugs with the same end result. Summarizing here to save newcomers having to read the thread

✅  Homebrew `zsh-abbr` formula had a bug [fixed. Read [#156 (comment)](https://github.com/olets/zsh-abbr/issues/156?show_full=true#issuecomment-2474349854) for how to apply the fix]
✅  Documentation for manual install needed updating [fixed. Read <https://zsh-abbr.olets.dev/installation.html#manual> for the updated documentation]
✅  GitHub-generated archives had a bug [fixed. Read <https://github.com/olets/zsh-abbr/releases/tag/v6.0.0> for how to get a complete archive]

Original report follows…

---
&nbsp;
&nbsp;

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Update the issue title

- [X] I have updated the title

### Expected Behavior

`source /opt/homebrew/share/zsh-abbr/zsh-abbr.zsh` should return no errors and "just work" ;-)

### Actual Behavior

```zsh
❯ brew install olets/tap/zsh-abbr
[ … no errors … ]
❯ source /opt/homebrew/share/zsh-abbr/zsh-abbr.zsh
abbr: There was a problem finishing installing dependencies
```

### Steps To Reproduce

1. macOS 14.7.1
2. Homebrew 4.4.5
3. zsh-abbr installed via `brew install olets/tap/zsh-abbr`
4. `source /opt/homebrew/share/zsh-abbr/zsh-abbr.zsh`

### Environment

```shell
❯ abbr profile
abbr:905: no such file or directory: /opt/homebrew/Cellar/zsh-abbr/6.0.0/share/zsh-abbr/zsh-job-queue/zsh-job-queue.zsh
abbr:906: no such file or directory: /opt/homebrew/Cellar/zsh-abbr/6.0.0/share/zsh-abbr/zsh-job-queue/zsh-job-queue.zsh
abbr:925: no such file or directory: /opt/homebrew/Cellar/zsh-abbr/6.0.0/share/zsh-abbr/zsh-job-queue/zsh-job-queue.zsh
zsh-abbr version 6.0.0
zsh 5.9 (x86_64-apple-darwin23.0)
OSTYPE darwin23.0
```

### Installation method

Homebrew

### Anything else?

Thanks for zsh-abbr, worked really great for me for the last years and makes working with ZSH so much more comfortable!

Please let me know if I can test so0mething or which additional info is needed – the error message itself is not very verbose …

**Comment by olets** [comment]
Thanks for reporting. Oversight in updating the formula to point to v6. Should be fixed.

This might work

```shell
brew update && brew reinstall zsh-abbr
```

This probably will

```shell
brew uninstall zsh-abbr && brew update && brew install zsh-abbr
```

This definitely should

```shell
brew uninstall zsh-abbr && brew update && brew install zsh-abbr
```

Lmk which if any of those work for you. Thanks!

**Comment by LangLangBart** [comment]
Hi, I followed the manual[^1] instructions and encountered this issue.

Below is a Docker command to consistently reproduce the issue.

```bash
docker run --rm --tty --interactive alpine:latest sh -uec '
apk add zsh git
git clone https://github.com/olets/zsh-abbr --single-branch --branch main --depth 1 /olet-abbr
echo "source /olet-abbr/zsh-abbr.plugin.zsh" > ~/.zshrc
/bin/zsh'
fetch https://dl-cdn.alpinelinux.org/alpine/v3.20/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.20/community/x86_64/APKINDEX.tar.gz
(1/18) Installing ca-certificates (20240705-r0)
(2/18) Installing brotli-libs (1.1.0-r2)
...
Resolving deltas: 100% (5/5), done.
abbr: Finishing installing dependencies
abbr: There was a problem finishing installing dependencies

8d860058da6f# abbr --version
abbr:905: no such file or directory: /olet-abbr/zsh-job-queue/zsh-job-queue.zsh
abbr:906: no such file or directory: /olet-abbr/zsh-job-queue/zsh-job-queue.zsh
abbr:925: no such file or directory: /olet-abbr/zsh-job-queue/zsh-job-queue.zsh
zsh-abbr version 6.0.0
```

[^1]: [Installation :: zsh-abbr](https://zsh-abbr.olets.dev/installation.html#manual)

**Comment by LangLangBart** [comment]
Using the `--recurse-submodules` flag with `git clone` does the trick. 

```txt
         --recurse-submodules[=<pathspec>]
             After the clone is created, initialize and clone submodules within
             based on the provided <pathspec>. If no =<pathspec> is provided,
             all submodules are initialized and cloned. This option can be given
             multiple times for pathspecs consisting of multiple entries. The
             resulting clone has submodule.active set to the provided pathspec,
             or "." (meaning all submodules) if no pathspec is provided.
   
             Submodules are initialized and cloned using their default settings.
             This is equivalent to running git submodule update --init
             --recursive <pathspec> immediately after the clone is finished.
             This option is ignored if the cloned repository does not have a
             worktree/checkout (i.e. if any of --no-checkout/-n, --bare, or
             --mirror is given)
```

---

The problem is that I use `zsh4humans`[^1] to install plugins, and there is no option I can pass to the installation process to update the submodules. 

```zsh
# auto-expanding abbreviations - https://github.com/olets/zsh-abbr
z4h install olets/zsh-abbr || return
```

[^1]: [GitHub - romkatv/zsh4humans: A turnkey configuration for Zsh](https://github.com/romkatv/zsh4humans)

**Comment by olets** [comment]
Thanks for the report @LangLangBart.

Make this change:

```diff
- git clone https://github.com/olets/zsh-abbr --single-branch --branch main --depth 1 /olet-abbr
+ git clone https://github.com/olets/zsh-abbr --recurse-submodules --single-branch --branch main --depth 1 /olet-abbr
```

Can fix an existing v6 clone with

```shell
cd path/to/zsh-abbr
git submodule update --init
```

I've updated the docs.

**Comment by olets** [comment]
Crossing comments!

Can you clone manually and then have z4h load that? If not, recommend opening a z4h issue to ask.

**Comment by LangLangBart** [comment]
Thanks @olets 

I switched to the `brew` install method, and it works well.

**Comment by llakala** [comment]
> The dependencies are also missing in the [source distribution files](https://github.com/olets/zsh-abbr/releases), will you add them there?

Bumping this, nixpkgs needs this before we can update to 6.0.0.

**Comment by bartlibert** [comment]
> > The dependencies are also missing in the [source distribution files](https://github.com/olets/zsh-abbr/releases), will you add them there?
> 
> Bumping this, nixpkgs needs this before we can update to 6.0.0.

FYI, for arch linux I switched the aur package to using git tags to work around this. Maybe it can help for nixos as well: [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=zsh-abbr)

**Comment by olets** [comment]
Interesting, didn't occur to me that GitHub's generated binaries might miss submodules. Seems like a bug.

Added. https://github.com/olets/zsh-abbr/releases/tag/v6.0.0

**Comment by olets** [comment]
@alexbarton @Chekote I believe your bugs are fixed

NixOS is still giving us trouble but imo it's fair to consider that a Nix package problem, and to discuss in https://github.com/NixOS/nixpkgs/pull/355512

Tentatively closing this. Comment if you're still having trouble

**Comment by arrrgi** [comment]
> Crossing comments!
> 
> Can you clone manually and then have z4h load that? If not, recommend opening a z4h issue to ask.

Raised an issue on zsh4humans.

**Comment by arrrgi** [comment]
@olets it might be too late now with this issue closed, but have you considered the benefits of using a subtree for including `zsh-job-queue`

I opened an issue with the **zsh4humans** maintainer and was told it wasn't trivial to provide (as per issue https://github.com/romkatv/zsh4humans/issues/335) as _z4h_ doesn't support installing Zsh plugins with a `git clone` operation.

**Comment by olets** [comment]
Never too late, but there's no question in my mind that submodule is the better tool for this case than subtree. I'll comment in the z4h with a workaround concept.

It's a bummer that the archives GitHub automatically generates for releases don't include submodules' contents (the root cause of most of the reports in this issue), but people whose go-to plugin management method doesn't support Git cloning or downloading user-specified archives can always use the docs' "manual" method (or in the case of zsh-abbr Homebrew if they're Homebrew users).

(Let's not make this the place to debate this subjective question, but for anyone reading along who hasn't used both (or either) subtree has its place but I don't find the above Atlassian article's premise or argument compelling. From a quick search, this is closer to how I think about the two https://adam-p.ca/blog/2022/02/git-submodule-subtree/.)

**Comment by romkatv** [comment]
> this is closer to how I think about the two https://adam-p.ca/blog/2022/02/git-submodule-subtree/

This article focuses on discussing pros and cons for the developer of the code. When it comes to the users of the code, submodules have nothing going for them, they are just pain.

**Comment by arrrgi** [comment]
I will accept Homebrew as the installation method as I don't think any further productive conversation will come of the debate between including source trees from other repos. I have used both submodules and subtrees and see merit in each approach.

I'm fortunate that the systems I develop on all have Homebrew installed before my Zsh config and personalisations are setup, but I would advocate that not all Linux users want to install Homebrew and would appreciate some form of automated way to include this in their ecosystem as a plugin for Zsh.

**Comment by olets** [comment]
> not all Linux users want to install Homebrew and would appreciate some form of automated way to include this in their ecosystem as a plugin for Zsh

For users who want to automate their own downloading of an archive but don't know where to start, and for users who\* can't/won't use a plugin manager that supports zsh-abbr, can't/won't use Homebrew, can't/won't use `git clone`, and aren't on a distro that has its own zsh-abbr package, I've added some GitHub REST API solution signposting to [the "installing as a plugin" docs](https://zsh-abbr.olets.dev/installation#plugin). Also tested 20 plugin managers, and documented limitations in the same section (tl;dr nearly all work with no special considerations needed).

> I'll comment in the z4h issue with a workaround concept

I've added a Git-dependent z4h postinstall hook snippet to [the docs' Integrations page](https://zsh-abbr.olets.dev/integrations). I'm not deeply familiar with z4h package management, and features may not be supported, but installing works.\*\* Also added signposting for a no-Git-dependency solution.

&nbsp;

\* Tone check: don't read this long list as snarky, just helping users determine whether they fall into this category
\*\* Iterating on that is outside the scope of this issue — open a new issue if necessary
